crates/interpreter/src/side_table.rs

```diff
@@ -14,26 +14,39 @@
 
 use alloc::vec::Vec;
 use core::ops::Range;
+use std::marker::PhantomData;
 
 use crate::error::*;
 use crate::module::Parser;
+use crate::toctou::*;
 
 pub const SECTION_NAME: &str = "wasefire-sidetable";
 
 #[derive(Debug)]
-pub struct SideTableView<'m> {
+pub struct SideTableView<'m, M: Mode> {
     pub indices: &'m [u8], // including 0 and the length of metadata_array
     pub metadata: &'m [u8],
+    pub mode: PhantomData<M>,
 }
 
-impl<'m> SideTableView<'m> {
-    // TODO(dev/fast-interp): Make it generic since it will be used in both `Check` and `Use` modes.
-    // (Returns `MResult<Metadata<'m>, M>` instead.)
-    pub fn metadata(&self, func_idx: usize) -> Metadata<'m> {
-        Metadata(
-            &self.metadata[parse_u16(self.indices, func_idx * 2) as usize
-                .. parse_u16(self.indices, (func_idx + 1) * 2) as usize],
-        )
+impl<'m, M: Mode> SideTableView<'m, M> {
+    pub fn metadata(&self, func_idx: usize) -> MResult<Metadata<'m>, M> {
+        try {
+            Metadata(
+                &self.metadata[parse_u16::<M>(
+                    self.indices,
+                    func_idx.checked_mul(2).ok_or_else(|| M::invalid())?,
+                ) as usize
+                    .. parse_u16::<M>(
+                        self.indices,
+                        func_idx
+                            .checked_add(1)
+                            .ok_or_else(|| M::invalid())?
```

Comment:
No need for ok_or_else here. This can just be `func_idx.checked_add(1)?.checked_mul(2)?`.

Also it's probably worth it to use some variables for the range start and limit.